abdm_integrator/hiu/views/health_information.py

```diff
@@ -0,0 +1,295 @@
+        if request_data.get('transaction_id') and request_data.get('page'):
+            page_number = request_data['page']
+            gateway_request_id = get_object_or_404(
+                HealthInformationRequest, transaction_id=request_data['transaction_id']).gateway_request_id
+        else:
+            page_number = 1
+            hiu_crypto = ABDMCrypto()
+            gateway_request_id = self.gateway_health_information_cm_request(artefact, health_info_url,
+                                                                            hiu_crypto.transfer_material)
+            self.save_health_info_request(request.user, artefact, gateway_request_id,
+                                          hiu_crypto.key_material.as_dict())
+
+        cache_key = f'{gateway_request_id}_{page_number}'
+        health_response_data = poll_for_data_in_cache(cache_key, interval=4)
+
+        self.handle_for_error(health_response_data)
+        if app_settings.HIU_PARSE_FHIR_BUNDLE:
+            health_response_data['entries'] = self.parse_fhir_bundle_for_ui(health_response_data['entries'])
+        response_data = self.generate_response_data(health_response_data, current_url,
+                                                    artefact.artefact_id)
+        return Response(status=HTTP_200_OK, data=response_data)
+
+    def validate_artefact_expiry(self, artefact):
+        if artefact.consent_request.expiry_date <= datetime.utcnow():
+            raise CustomError(
+                error_code=HIUError.CODE_CONSENT_EXPIRED,
+                error_message=HIUError.CUSTOM_ERRORS[HIUError.CODE_CONSENT_EXPIRED],
+            )
+
+    def gateway_health_information_cm_request(self, artefact, health_info_url, hiu_transfer_material):
+        payload = ABDMRequestHelper.common_request_data()
+        payload['hiRequest'] = {'consent': {'id': str(artefact.artefact_id)}}
+        payload['hiRequest']['dateRange'] = artefact.details['permission']['dateRange']
+        payload['hiRequest']['dataPushUrl'] = health_info_url
+        payload['hiRequest']['keyMaterial'] = hiu_transfer_material
+        logging.info(f"HIU: Request Health Information {payload}")
+        ABDMRequestHelper().gateway_post(HIUGatewayAPIPath.HEALTH_INFO_REQUEST, payload)
+        return payload['requestId']
+
+    def save_health_info_request(self, user, artefact, gateway_request_id, key_material_dict):
+        HealthInformationRequest.objects.create(user=user, consent_artefact=artefact,
+                                                gateway_request_id=gateway_request_id,
+                                                key_material=key_material_dict)
+
+    def _get_next_query_params(self, response_data, artefact_id):
+        params = QueryDict('', mutable=True)
+        params['artefact_id'] = artefact_id
+        params['transaction_id'] = response_data['transactionId']
+        params['page'] = response_data['pageNumber'] + 1
```

Comment:
Consider doing similar to https://github.com/dimagi/abdm-python-integrator/pull/5/files#r1381322510 to be DRY

Reply:
Addressed in ce8722ad9a202048ac5f6c11a52043ae96db3fbd